### ISSUE TITLE
Help needed: Volume setting value synchronization back to iOS device

### ISSUE THREAD

**Issue by PMatThunderData** [mikebrady/shairport-sync#1215]
Hi Mike,

We are developing a product that has software volume control. And when the user use a iOS device to play music through AirPlay to play music with the product, he/she can use another volume control feature (a knob or another mobile app) to change the volume, we need to notify the iOS device to synchronize the volume setting. However, we failed to find a relative function in shairport-sync source code to make it. Could you advise how we can make it?

Thank you very much!

**Comment by mikebrady** [comment]
Thanks for the post. Have a look at the D-Bus interface. An approach would be for the "...other control feature" to send messages to iOS and for it, in turn, to adjust the volume. But please remember, this is not commercial-grade software!

**Comment by PMatThunderData** [comment]
Thank you for this information! We will try it asap.
And thank you for reminding, we will sign for MFi afterward.

**Comment by PMatThunderData** [comment]
Hi Mike,

We are now blocked at the part that it seems iOS device do not get the updated volume value from RTSP protocol.
in rtsp.c , there is a GET_PARAMETER, however we got failed to let iOS device to get the updated volume value.
Is there anything else we should take care?

Thank you!

**Comment by mikebrady** [comment]
@PMatThunderData, I do not think you can use the `GET_PARAMETER` facility to send the updated volume to the iOS device.

The approach I was suggesting -- the only one that I know may work -- was to use the ability to send remote control commands to the iOS device.

Shairport Sync has an optional D-Bus interface (use the `--with-dbus-interface` when building Shairport Sync). Have a look in the `sample bus commands` file in the `Documents` folder for some examples of using the CLI command `dbus-send` that allows you to send D-Bus messages from the command line.

One of those D-Bus commands allows you to request the iOS device to set the volume. The way it works is that  Shairport Sync passes this request as a remote control command to the iOS device.

 For example:

```
dbus-send --system --print-reply --type=method_call \
    --dest=org.gnome.ShairportSync '/org/gnome/ShairportSync' \
    org.gnome.ShairportSync.RemoteControl.SetAirplayVolume double:-10.0
``` 

However, while iOS will respond to remote control commands, not all devices or player will respond -- some third-party players do not.

(BTW, Shairport Sync must be _fully installed_ for the D-Bus interface to work.)

**Comment by PMatThunderData** [comment]
Hi Mike,

Thank you for detail information. We have go thorough and working on the detail implementation and polishing the code.

**Comment by PMatThunderData** [comment]
Hi Mike,

I am sorry that we need your help again. We found iPhone X with iOS 14.6 will not update the volume setting when the volume was changed. However, iMac with Apple Music will update the volume setting smoothly.

What below is the command that we do dbus-send. The $volume is the volume setting.

`dbus-send --system --print-reply --type=method_call --dest=org.gnome.ShairportSync '/org/gnome/ShairportSync' org.gnome.ShairportSync.AdvancedRemoteControl.SetVolume int32:$volume`
